tools/kvmexit.py

```diff
@@ -0,0 +1,293 @@
+    "EXIT_REASON_RDTSC",
+    "",
+    "EXIT_REASON_VMCALL",
+    "EXIT_REASON_VMCLEAR",
+    "EXIT_REASON_VMLAUNCH",
+    "EXIT_REASON_VMPTRLD",
+    "EXIT_REASON_VMPTRST",
+    "EXIT_REASON_VMREAD",
+    "EXIT_REASON_VMRESUME",
+    "EXIT_REASON_VMWRITE",
+    "EXIT_REASON_VMOFF",
+    "EXIT_REASON_VMON",
+    "EXIT_REASON_CR_ACCESS",
+    "EXIT_REASON_DR_ACCESS",
+    "EXIT_REASON_IO_INSTRUCTION",
+    "EXIT_REASON_MSR_READ",
+    "EXIT_REASON_MSR_WRITE",
+    "EXIT_REASON_INVALID_STATE",
+    "EXIT_REASON_MSR_LOAD_FAIL",
+    "",
+    "EXIT_REASON_MWAIT_INSTRUCTION",
+    "EXIT_REASON_MONITOR_TRAP_FLAG",
+    "",
+    "EXIT_REASON_MONITOR_INSTRUCTION",
+    "EXIT_REASON_PAUSE_INSTRUCTION",
+    "EXIT_REASON_MCE_DURING_VMENTRY",
+    "",
+    "EXIT_REASON_TPR_BELOW_THRESHOLD",
+    "EXIT_REASON_APIC_ACCESS",
+    "EXIT_REASON_EOI_INDUCED",
+    "EXIT_REASON_GDTR_IDTR",
+    "EXIT_REASON_LDTR_TR",
+    "EXIT_REASON_EPT_VIOLATION",
+    "EXIT_REASON_EPT_MISCONFIG",
+    "EXIT_REASON_INVEPT",
+    "EXIT_REASON_RDTSCP",
+    "EXIT_REASON_PREEMPTION_TIMER",
+    "EXIT_REASON_INVVPID",
+    "EXIT_REASON_WBINVD",
+    "EXIT_REASON_XSETBV",
+    "EXIT_REASON_APIC_WRITE",
+    "EXIT_REASON_RDRAND",
+    "EXIT_REASON_INVPCID",
+    "EXIT_REASON_VMFUNC",
+    "EXIT_REASON_ENCLS",
+    "EXIT_REASON_RDSEED",
+    "EXIT_REASON_PML_FULL",
+    "EXIT_REASON_XSAVES",
+    "EXIT_REASON_XRSTORS",
+    "",
+    "",
+    "EXIT_REASON_UMWAIT",
+    "EXIT_REASON_TPAUSE"
+)
+
+try:
+    has_kvm = subprocess.check_output("lsmod | grep kvm", shell=True)
+except Exception as e:
+    raise Exception("Please insmod kvm module to use kvmexit")
+
+try:
+    uname = subprocess.check_output("uname -r", shell=True).split('.')[:2]
+    if int(uname[0]) >= 5:
+        func_entry = "RAW_TRACEPOINT_PROBE(kvm_exit)"
+        get_er = "ctx->args[0]"
+    elif uname[0] == "4":
+        if int(uname[1]) >= 7:
+            func_entry = "TRACEPOINT_PROBE(kvm, kvm_exit)"
+            get_er = "args->exit_reason"
+        else:
+            raise Exception("Kernel version < 4.7 is not supported")
+    else:
+        raise Exception("Only kernel version >= 4.7 is supported")
+except Exception as e:
+    raise Exception("Failed to run kvmexit due to: %s" % e)
+
+
+# For kernel >= 5.0, use RAW_TRACEPOINT for performance consideration
+bpf_text = bpf_text.replace('FUNC_ENTRY', func_entry)
+bpf_text = bpf_text.replace('GET_ER', get_er)
+b = BPF(text=bpf_text)
+
+
+# header
+print("Display kvm exit reasons and statistics", end="")
+if duration < 99999999:
+    print(" for %d secs." % duration)
+else:
+    print("... Hit Ctrl-C to end.")
+print("%-8s %-8s %-35s %s" % ("TGID", "PID", "KVM_EXIT_REASON", "STAT"))
+
+# signal handler
+def signal_ignore(signal, frame):
+        print()
+try:
+    sleep(duration)
+except KeyboardInterrupt:
+    signal.signal(signal.SIGINT, signal_ignore)
+
+# output
+pcpu_kvm_stat = b["pcpu_kvm_stat"]
+pcpu_cache = b["pcpu_cache"]
+EXIT_REASON_NUM = 69
```

Comment:
Why hard code this, use `len(exit_reasons)` instead.

Reply:
> Use `len(exit_reasons)` instead please.

Thanks for the reminder.